compatibility_spark-3/src/main/scala/za/co/absa/hyperdrive/compatibility/impl/DeltaCDCToSnapshotWriter.scala

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.compatibility.impl
+
+import za.co.absa.hyperdrive.compatibility.api.{CompatibleDeltaCDCToSnapshotWriter, DeltaCDCToSnapshotWriterConfiguration}
+import io.delta.tables.{DeltaMergeBuilder, DeltaTable}
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.spark.sql.{Column, DataFrame, Row, SaveMode, SparkSession, functions}
+import org.apache.spark.sql.catalyst.expressions.objects.AssertNotNull
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.streaming.{OutputMode, StreamingQuery}
+import org.slf4j.LoggerFactory
+
+import java.net.URI
+
+class DeltaCDCToSnapshotWriter(configuration: DeltaCDCToSnapshotWriterConfiguration) extends CompatibleDeltaCDCToSnapshotWriter {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val STRING_SEPARATOR = "#$@"
+  private val CHECKPOINT_LOCATION = "checkpointLocation"
+
+  if(configuration.precombineColumnsCustomOrder.values.flatten.toSeq.contains(STRING_SEPARATOR)) {
+    throw new IllegalArgumentException(s"Precombine columns custom order cannot contain string separator: $STRING_SEPARATOR")
+  }
+
+  override def write(dataFrame: DataFrame): StreamingQuery = {
+    dataFrame.writeStream
+      .trigger(configuration.trigger)
+      .outputMode(OutputMode.Append())
+      .option(CHECKPOINT_LOCATION, configuration.checkpointLocation)
+      .options(configuration.extraConfOptions)
+      .foreachBatch((df: DataFrame, batchId: Long) => {
+        if(!DeltaTable.isDeltaTable(df.sparkSession, configuration.destination)) {
+          if(isDirEmptyOrDoesNotExist(df.sparkSession, configuration.destination)) {
+            logger.info(s"Destination: ${configuration.destination} is not a delta table. Creating new delta table.")
+            df.sparkSession
+              .createDataFrame(df.sparkSession.sparkContext.emptyRDD[Row], df.schema)
+              .write
+              .format("delta")
+              .mode(SaveMode.Overwrite)
+              .option("overwriteSchema", "true")
+              .partitionBy(configuration.partitionColumns :_*)
+              .save(configuration.destination)
+          } else {
+            throw new IllegalArgumentException(s"Could not create new delta table. Directory ${configuration.destination} is not empty!")
```

Comment:
Exception is thrown when you configure the checkpoints folder within the data directory. Actually it doesn't make sense to do this check in every microbatch, but it should rather happen before the query runs. So I would move the logic to the start of the method.

Reply:
Done